src/include/optrack.h

```diff
@@ -0,0 +1,67 @@
+/*-
+ * Copyright (c) 2014-2017 MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#define	WT_OPTRACK_MAXRECS 16384
+#define	WT_OPTRACK_BUFSIZE WT_OPTRACK_MAXRECS * sizeof(WT_TRACK_RECORD)
+
+/*
+ * WT_TRACK_RECORD --
+ *     A structure for logging potentially long operations.
+ *
+ * We pad the record so that the size of the entire record is three double
+ * words, or 24 bytes. If we don't do this, the compiler will pad it for us,
+ * because we keep records in the record buffer array and each new record must
+ * be aligned on the 8-byte boundary, since its first element is an 8-byte
+ * timestamp. Instead of letting the compiler insert the padding silently, we
+ * pad explicitly, so that whoever writes the binary decoder can refer to this
+ * struct to find out the record size.
+ */
+struct __wt_track_record {
+	uint64_t timestamp;
+	uint64_t op_id;
+	uint16_t op_type;
+	char padding[6];
+};
+
+#define	WT_TRACK_OP(s, optype)						\
+	tr->timestamp = __wt_rdtsc(s);					\
+	tr->op_type = optype;						\
+	tr->op_id = (uint64_t)&__func__;				\
+									\
+	if (optype == 0 && !id_recorded)				\
+		__wt_optrack_record_funcid(s, tr->op_id,		\
+		    (void*)&__func__, sizeof(__func__), &id_recorded);	\
+	if (s->optrackbuf_ptr == WT_OPTRACK_MAXRECS) {			\
+		if (s->optrack_fh != NULL)				\
```

Comment:
I doubt this condition will ever be true. __wt_optrack_flush_buffer() allocates the optrack_fh, and __wt_optrack_flush_buffer() gets called either at session close or here. And since optrack_fh will be NULL here, it will not call __wt_optrack_flush_buffer().

Reply:
Good catch! We used to allocate optrack_fh on session open, so it would typically not be NULL here. Then I changed to lazy allocation of operation tracking files. Now it will always be NULL. I will fix the logic.